source/tutorials/file-sets.md

```diff
@@ -0,0 +1,661 @@
+
+```shell-session
+nix-repl> fs.trace ./. null
+trace: /home/user (all files in directory)
+null
+```
+
+All functions that expect a file set for an argument also accept a [path](https://nixos.org/manual/nix/stable/language/values#type-path) instead.
+Such path arguments are then [implicitly turned into sets](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset-path-coercion) that contain _all_ files under the given path.
+In the previous trace this is indicated by `(all files in directory)`.
+
+:::{tip}
+The `trace` function pretty-prints its first argument and returns its second argument.
+But since you often just need the pretty-printing in `nix repl`, you can omit the second argument:
+
+```shell-session
+nix-repl> fs.trace ./.
+trace: /home/user (all files in directory)
+«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
+```
+:::
+
+Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
+So You don't have to worry as much about accidentally copying secrets into the world-readable store.
+
+In this example, although we pretty-printed the home directory, no files were copied.
+This is in contrast to coercion of paths to strings such as in `"${./.}"`,
+which copies the whole directory to the Nix store on evaluation!
+
+:::{warning}
+When using [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), local directories with a `flake.nix` are always copied into the Nix store *completely* unless it is a Git repository!
+:::
+
+This implicit coercion also works for files:
+
+```shell-session
+$ touch some-file
+```
+
+```shell-session
+nix-repl> fs.trace ./some-file
+trace: /home/user
+trace: - some-file (regular)
+```
+
+In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).
+
+
+## Example project
+
+To further experiment with the library, make a sample project.
+Create a new directory, enter it,
+and set up `niv` to pin the Nixpkgs dependency:
+
+```shell-session
+$ mkdir fileset
+$ cd fileset
+$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch master"
+```
+
+Then create a `default.nix` file with the following contents:
+
+```{code-block} nix
+:caption: default.nix
+{
+  system ? builtins.currentSystem,
+  sources ? import ./nix/sources.nix,
+}:
+let
+  pkgs = import sources.nixpkgs {
+    config = { };
+    overlays = [ ];
+    inherit system;
+  };
+in
+pkgs.callPackage ./build.nix { }
+```
+
+Add two source files to work with:
+
+```shell-session
+$ echo hello > hello.txt
+$ echo world > world.txt
+```
+
+## Adding files to the Nix store
+
+To add files in a given file set to the Nix store,
+we use [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).
```

Comment:
Who is "we" here? It makes sense to say "we" when we're referring to this group writing the tutorial, e.g. when explaining why we're telling people to fetch from `master` rather than `nixpkgs-unstable`. But here here we'd run the danger of saying "we use this, you do you", while we really want to either state a fact "X does Y" or tell people what to do.
```suggestion
Files in a given file set can be added to the Nix store with [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).
```

I don't super like the phrasing, but it should be good enough.